docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+- Enable more precise type definitions that can capture complex relationships and constraints
+- Facilitate the creation of reusable and composable libraries
+- Enhance code safety and maintainability with better type-level programming
+
+The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.
+
+## Design
+
+Type functions wil be defined with the following syntax:
+```luau
+type function f(...)
+    -- implementation of the type function
+end
+```
+
+For instance, the `rawget` type function can be written as:
+```luau
+type function rawget(tbl, prop)
+    return tbl.getProps()[prop.getValue()]
+
+    -- returning nothing / nil means that this type function will fail to reduce
+end
+
+type Person = {
+    name: string,
+    age: number
+}
+
+type ty = rawget<Person, "name"> -- ty = string
+```
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
+
+<details><summary>lType library methods (dropdown)</summary>
+Note: methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types (e.g, for `getIndexer()`, assert that `isTable()` is true).
+
+#### Any
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `isString()` | `boolean` | returns true if self is of type `string` |
+| `isNumber()` | `boolean` | returns true if self is of type `number` |
+| `isBoolean()` | `boolean` | returns true if self is of type `boolean` |
+| `isTable()` | `boolean` | returns true if self is of type `table` |
+| `isThread()` | `boolean` | returns true if self is of type `thread` |
+| `isFunction()` | `boolean` | returns true if self is of type `function` |
+| `isBuffer()` | `boolean` | returns true if self is of type `buffer` |
+| `isNil()` | `boolean` | returns true if self is of type `nil` |
+| `isClass()` | `boolean` | returns true if self is of type `class` (do we need this?) |
+| `isBooleanSingleton()` | `boolean` | returns true if self is a boolean singleton |
+| `isStringSingleton()` | `boolean` | returns true if self is a string singleton |
+| `isA(arg: lType)` | `boolean` | returns true if arg is the same type as self |
```

Comment:
What motivated this one?

Reply:
the motivation behind `isa()` was to provide a way for developers to compare the relationships of different types. I saw [your comment](https://github.com/luau-lang/rfcs/pull/45#discussion_r1651707952) and thought that it was a better idea than what I proposed, so I am planning on making the change to your idea on my revision.